core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UseConnectorTableScanPartitioning.java

```diff
@@ -56,11 +61,23 @@ public Result apply(TableScanNode node, Captures captures, Context context)
         }
 
         TablePartitioning partitioning = properties.getTablePartitioning().get();
-        if (nodePartitioningManager.getConnectorBucketNodeMap(context.getSession(), partitioning.getPartitioningHandle()).hasFixedMapping()) {
+        ConnectorBucketNodeMap bucketNodeMap = nodePartitioningManager.getConnectorBucketNodeMap(context.getSession(), partitioning.getPartitioningHandle());
+        if (bucketNodeMap.hasFixedMapping()) {
             // use connector table scan partitioning when bucket to node assignments are fixed
             return Result.ofPlanNode(node.withUseConnectorTablePartitioning(true));
         }
 
-        return Result.ofPlanNode(node.withUseConnectorTablePartitioning(isPlanWithTableNodePartitioning(context.getSession())));
+        if (!isPlanWithTableNodePartitioning(context.getSession())) {
+            return Result.ofPlanNode(node.withUseConnectorTablePartitioning(false));
+        }
+
+        int numberOfBuckets = bucketNodeMap.getBucketCount();
+        int numberOfTasks = taskCountEstimator.estimateSourceDistributedTaskCount(context.getSession());
+
+        if ((double) numberOfBuckets / numberOfTasks < getTableScanPartitioningMinBucketToTaskRatio(context.getSession())) {
+            return Result.ofPlanNode(node.withUseConnectorTablePartitioning(false));
```

Comment:
Should we add a `Warning` ?

Reply:
I don't think so. This is expected situation and it should improve wall time. It would be similar to adding warning for CBO decisions.